src/auto-evo/EvolutionaryTree.cs

```diff
@@ -226,18 +293,216 @@ private void BuildTree(uint id, ref uint index)
+                dirty = true;
+            }
+        }
+        else if (@event is InputEventMouseMotion motionEvent)
+        {
+            if (dragging)
+            {
+                var delta = (motionEvent.Position - lastMousePosition) / sizeFactor;
+                dragOffset += horizontalOnly ? new Vector2(delta.x, 0) : delta;
+                lastMousePosition = motionEvent.Position;
+                BindOffsetToTreeSize();
+                dirty = true;
+            }
+        }
+    }
+
+    private void BindOffsetToTreeSize()
+    {
+        // TreeSize may be less than RectSize, so the later Min and Max is not merged into Clamp.
+        // Note that dragOffset's x and y should both be negative.
+        var start = RectSize / sizeFactor - TreeSize;
+
+        float x = dragOffset.x;
+        x = Math.Max(x, start.x);
+        x = Math.Min(x, 0);
+
+        float y = dragOffset.y;
+        y = Math.Max(y, start.y);
+        y = Math.Min(y, 0);
+
+        dragOffset = new Vector2(x, y);
+    }
+
+    /// <summary>
+    ///   Stop dragging when mouse exit <see cref="timeline"/> or <see cref="tree"/>
+    /// </summary>
+    private void MouseExit()
+    {
+        dragging = false;
+    }
+
+    /// <summary>
+    ///   <see cref="Control._Draw"/> of <see cref="tree"/>
+    /// </summary>
+    private void TreeDraw()
+    {
+        var drawRegion = new Rect2(RectPosition - DrawMargin, RectSize + DrawMargin);
+
+        // Draw new species connection lines
+        foreach (var node in speciesNodes.Values.Select(l => l.First()))
+        {
+            if (node.ParentNode == null)
+                continue;
+
+            // If the vertical position is outside draw region, skip.
+            if (node.Center.y < drawRegion.Position.y || node.ParentNode.Center.y > drawRegion.End.y)
+                continue;
+
+            // Horizontal too.
+            if (node.Center.x < drawRegion.Position.x || node.ParentNode.Center.x > drawRegion.End.x)
+                continue;
+
+            TreeDrawLine(node.ParentNode.Center, node.Center);
+        }
+
+        float treeRightPosition =
+            sizeFactor * (dragOffset.x + GENERATION_SEPARATION * latestGeneration + TreeNodeSize.x);
+
+        // Draw horizontal lines
+        foreach (var nodeList in speciesNodes.Values)
+        {
+            var lineStart = nodeList.First().Center;
+
+            // If the line is outside draw region, skip it.
+            if (lineStart.y < drawRegion.Position.y || lineStart.y > drawRegion.End.y)
+                continue;
+
+            var lastNode = nodeList.Last();
+
+            // If species extinct, line ends at the last node; else it ends at the right end of the tree
+            var lineEnd = lastNode.LastGeneration ?
+                lastNode.Center :
+                new Vector2(treeRightPosition, lineStart.y);
+
+            TreeDrawLine(lineStart, lineEnd);
+        }
+
+        latoSmallItalic.Set("size", sizeFactor * SMALL_FONT_SIZE);
```

Comment:
Here's the font size, so I assume with a similar call in _Ready this could get the default font size automatically and not have to assume it is 14.